### PR TITLE
drivers: i3c: add additional variable for num xfer for i3c

### LINKED ISSUE #65099: I3C: reporting earlier than expected EoD or CCC with different sizes
# Description
With I3C, when a controller does a read whether, through a private transfer or a CCC, the target is responsible for issuing the End of Data through the T bit, but the way the API is currently implemented, there is no way to communicate this back up. For example, if a controller tries to read 100 Bytes (where the controller is to do an abort after 100 (or 101 bytes)) from a Target and the Target gives the EoD at the 42nd Byte, then there is no way to currently know where the EoD occurred and how far it can be read up to in the provided buffers.

This can also present a problem when trying to perform a CCC MXDS which the format used is defined based on the number of bytes returned by the Target. There needs to be a way for the length to be communicated back up.

# Proposed Solution
One Solution would be for the driver to overwrite the data length of the packet passed to it for the application then to read and know if the data ended earlier than expected, but this just ""feels"" kinda dirty. As the application would always need to store another copy of the data length it just first passed into the API for it to compare if it was modified after it returns.
Another solution would be an additional variable to be added within `i3c_msg` and 'other's' which would be actual data sent/received which the driver would be expected to write to.

### PR DESCRIPTION
**Pull request by XenuIsWatching**
With I3C, when a controller does a read whether, through a private transfer
or a CCC, the target is responsible for issuing the End of Data through the
T bit, but the way the API is currently implemented, there is no way to
communicate this back up. For example, if a controller tries to read 100
Bytes (where the controller is to do an abort after 100 (or 101 bytes))
from a Target and the Target gives the EoD at the 42nd Byte, then there is
no way to currently know where the EoD occurred and how far it can be read
up to in the provided buffers.
    
This implements a num_xfer for `i3c_msg` and `i3c_ccc_payload` which the
driver is responsible for writing the total number of bytes transfered.

Fixes #65099

### CODE REVIEW

File: drivers/i3c/i3c_cdns.c [hunk 1]
```diff
@@ -1732,6 +1739,8 @@ static int cdns_i3c_transfer(const struct device *dev, struct i3c_device_desc *t
 		} else {
 			send_broadcast = true;
 		}
+
+		cmd->num_xfer = msgs[i].num_xfer;
```
Comment (dcpleung) [inline drivers/i3c/i3c_cdns.c:1743]: Shouldn't this be the other way around? `msgs[i].num_xfer = cmd->num_xfer`?

File: drivers/i3c/i3c_mcux.c [hunk 2]
```diff
@@ -1131,6 +1131,9 @@ static int mcux_i3c_transfer(const struct device *dev,
 			goto out_xfer_i3c_stop_unlock;
 		}
 
+		/* write back the total number of bytes transferred */
+		*msgs[i].num_xfer = ret;
```
Comment (dcpleung) [inline drivers/i3c/i3c_mcux.c:1135]: This is not a pointer so there is no need for the `*` beforehand.

File: drivers/i3c/i3c_cdns.c (hunk 1 above)
Comment (XenuIsWatching) [reply drivers/i3c/i3c_cdns.c:1743]: no, this is just writing the pointer for reference, which is then updated within `cdns_i3c_complete_transfer`

File: drivers/i3c/i3c_mcux.c (hunk 2 above)
Comment (XenuIsWatching) [reply drivers/i3c/i3c_mcux.c:1135]: uh... that is a pointer

Comment (XenuIsWatching) [reply drivers/i3c/i3c_mcux.c:1135]: would you prefer it to not be a pointer?

Comment (XenuIsWatching) [reply drivers/i3c/i3c_mcux.c:1135]: if this stays a pointer... I probably should of checked it if it is NULL :(

Comment (dcpleung) [reply drivers/i3c/i3c_mcux.c:1135]: Ah... never mind... I thought `num_xfer` was just an integer.

Though, I think it is probably better for the ones in the structs to _not_ be pointers. The users of these APIs will not need to keep another variable to store this data as it is already in the `i3c_msg` struct.

Comment (XenuIsWatching) [reply drivers/i3c/i3c_mcux.c]: > Ah... never mind... I thought `num_xfer` was just an integer.
> 
> Though, I think it is probably better for the ones in the structs to _not_ be pointers. The users of these APIs will not need to keep another variable to store this data as it is already in the `i3c_msg` struct.

Yeah, the more I think about it. I'd have to agree... PR updated

**Review by dcpleung — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Review by decsny — APPROVED** [review]

**Pull request merged**